### ISSUE TITLE
Intermittent hangs with nested futures

### ISSUE THREAD

**Issue by mastfish** [ruby-concurrency/concurrent-ruby#369]
Code like
```
Concurrent::Future.execute{
 Concurrent::Future.execute{puts "pretend real code is here"}
}
```
Will occasionally cause the thread to hang forever. This is very intermittent (maybe once per hundred calls), and I've only been able to see it in logs from our production server. We're running MRI, 2.2.2, with version 0.9.0 of this gem.

I'm not sure if this is a limitation inherent to futures in ruby (perhaps something like dataflow should be used instead?), or if this is a bug.

**Comment by pitr-ch** [comment]
@mastfish Thank you for reporting, could you let us know how did you solve the issue?

**Comment by mastfish** [comment]
@pitr-ch I _think_ this wasn't caused by the futures, but by the network call we were doing inside those futures. Still gathering information though, I'll re-open and give more details if the source turns out to be futures.

**Comment by pitr-ch** [comment]
@mastfish thank you!

**Comment by jdantonio** [comment]
@mastfish Thank you for reporting this and looking into the possible cause. Definitely open the issue again if you continue to experience issues.

Internally `Future` is just a wrapper around the global thread pool. Although I've never tried to use nested futures, off the top of my head I can't think of why this would cause a thread to hang indefinitely. The internal future should simply add a new job to the queue then exit.

**Comment by mastfish** [comment]
The scenario I was thinking of was that all of the working threads handling the outer futures would be blocking on the futures inside, which would mean the futures inside would never get any processor time (or perhaps be blocked by the GIL). I'm not sure whether this could happen on MRI, though. It also seems like you'd need enough nested futures that all threads in the pool would be busy.

In any case, we tracked our issue to a misbehaving network call, so this potential problem is presumably theoretical (if it exists) rather than practical.

**Comment by pitr-ch** [comment]
If there are IO operations in futures, and there is no throttling system, the application can indeed run out of threads under heavy workload. The IO pool tries to prevent this by creating new threads if there are resources to create new ones, after that it queues the actions which brings danger of deadlocking. Based on my experience, it's best to enforce a limit on concurrently running IO operations.

**Comment by pitr-ch** [comment]
You can also try edge futures, which are slightly faster and they utilise less threads. The underlying danger is still there though.

**Comment by jdantonio** [comment]
@mastfish The thread pools also have a queue, however. The `#execute` method returns when the task is added to the pool, *not* once a thread picks up the task. So even is a pool is starved, a nested `Future` shouldn't cause a deadlock. The inner `#execute` will simply queue up the job and exit, returning control to the outer `#execute` call.

It is possible to create a thread pool with a limited queue size, but our global thread pools do not do this. Even when the pool's queue size is limited and the queue is full, a deadlock should not occur. The `fallback_policy` of the queue thread pool will be executed instead. The default option is to raise an exception. The other options are to run the operation on the current thread or to discard the task. All of those options should prevent a deadlock.

**Comment by pitr-ch** [comment]
As @jdantonio said, we try to do as much as possible to prevent deadlocks, but there still can be case when the futures task's completion is dependent on completion of other futures. Let's consider following scenario:
- application is handling concurrently incoming request
- each request results in two futures A and B where A depends on B and B is created some time after A.
- if the application is flooded with requests it uses all available threads with A tasks, rest is going to queue including B tasks. 
- all threads are occupied/blocked by A tasks waiting on B tasks in queue.
- result is deadlock :( 

@jdantonio We should probably write something about this to README or somewhere else to prevent our users being bitten by this.

**Comment by ankuriitk** [comment]
Reigniting this old thread. Is there a recommendation for the nesting problem? Is the solution that rate limit Future As so that Future B always have at one thread free available from the global pool? Another idea that came to mind was have dedicated thread pools for each level of nesting  (what if B spawns Future C and B depends on value of C). Is there any other alternative to avoid the deadlock?

**Comment by pitr-ch** [comment]
There is a solution in edge promises (still in a PR) which will be merged after 1.0.3 release and then made available in edge. In the meantime I recommend to control how many futures (or promises) of A is created.

I am not sure about the nesting pools. It creates unnecessary complexity and also does not fixes the underlying problem. It can only postpone the problem if the tasks are laid appropriately to the nesting.

In general value method should be avoided whenever possible, because it blocks and as you noted may lead to deadlocks. Chaining of promises does not block a thread and should be preferred. (Nesting does not block threads in edge promises). JS does not even have the value method and users are forced to use then.
